src/promnesia/sources/viber_desktop.py

```diff
@@ -0,0 +1,164 @@
+
+    # TODO for displaying maybe better not to expand/absolute, 
+    # but it's safer for debugging resolved.
+    path = Path(database).expanduser().resolve().absolute()
+    assert path.is_file(), path
+
+    # TODO context manager?
+    db = dataset_readonly(path)  # TODO could check is_file inside
+
+    query_str = textwrap.dedent(
+        f"""
+        /*
+        Establish group-names by concatenating:
+        - groups explicitely named,
+        - multi-groups having a group-leader (PGRole=2), with
+        - all the rest groups that (must) have just 2 members,
+        me(ContactId=1) & "other" contact, so use this as group-name.
+        */
+        WITH G0 AS (
+            SELECT
+                CR.ChatId,
+                CR.ContactID,
+                coalesce(CI.Name, C.Name, C.ClientName) as chatname,
+                CR.PGRole,
+                CI.PGTags
+            FROM ChatRelation as CR
+            JOIN Contact AS C ON CR.ContactID = C.ContactID
+            JOIN ChatInfo as CI ON CR.ChatId = CI.ChatId
+        ), G1 AS (
+            SELECT * FROM G0 WHERE PGRole = 2
+        ), G2 AS (
+            SELECT * FROM G0 WHERE
+                ContactID <> 1 AND
+                ChatId NOT IN (SELECT ChatId FROM G1)
+        ), Groups AS (
+            SELECT ChatId, chatname, PGTags FROM G1
+            UNION
+            SELECT ChatId, chatname, PGTags FROM G2
+        )
+        SELECT
+            M.EventId           AS mid,
+            E.TimeStamp         AS time,
+            G.chatname          AS chatname,
+            coalesce(
+                S.Name,
+                S.ClientName,
+                '(' || S.Number || ')'          /* contacts have one xor the other, but failsafe */
+            )                   AS sender,
+            coalesce(M.Subject, M.Body)         /* didn't see any msg with both */    
+                                AS text,
+            M.info              AS infojson,    /* to harvested titles from embedded urls */
+            G.PGTags            AS tags
+        FROM messages AS M
+        LEFT JOIN Events AS E
+            ON M.EventId = E.EventId
+        LEFT JOIN Contact AS S
+            ON E.ContactId = S.ContactId
+        LEFT JOIN Groups AS G
+            ON E.ChatId = G.ChatId
+        WHERE
+            text LIKE '%http%'
+        ORDER BY time;
+        """
+    )
+
+    def _parse_json_title(js) -> str:
+        if js and js.strip():
+            js = json.loads(js)
+            if isinstance(js, dict):
+                return js.get("Title")
+
+    def _handle_row(row) -> Results:
+        text = row["text"]
+        assert (
+            text
+        ), f"sql-query should have eliminated messages not containing 'http': {text}"
+        urls = extract_urls(text)
+        if not urls:
+            return
+        dt = from_epoch(row["time"] // 1000)  # timestamps are stored x100 this db
+        mid: str = unwrap(row["mid"])
```

Comment:
Oh! my bad, for `unwrap` in telegram module I intented to do `notnone` instead, [like in HPI](https://github.com/karlicoss/HPI/blob/ad924ebca84b0846c98401b10e61488c45cf1e9c/my/core/error.py#L18-L20). The reason was to make sure null rows aren't converted to `None` in python and avoid downstream errors.

`Exception` can never be in the sqlite row, so you can either remove `unwrap` from this module, or use `notnone` instead, up to you.

Reply:
Resolved in a8c6dd0.